File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
    int numDistinct(string& s, string& t, int i, int j, vector<vector<int>>& dp) {
        
        if(j == t.size())
            return 1;
        if(i == s.size())
            return 0;
           
        if(dp[i][j] != -1)
            return dp[i][j]; 
        
        if(s[i] != t[j])
            return dp[i][j] = numDistinct(s, t, i+1, j, dp);
        
        return dp[i][j] = numDistinct(s, t, i+1, j+1, dp) + numDistinct(s, t, i+1, j, dp);
        
    }
public:
    int numDistinct(string s, string t) {
        vector<vector<int>>dp(s.size(), vector<int>(t.size(), -1));
        return numDistinct(s, t, 0, 0, dp);
    }
};

File: 0115-distinct-subsequences/0115-distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        
        dp = {}
        M, N = len(s), len(t)
        
        def numDistinct(i, j):
            
            if j == N:
                return 1
            if i == M or M - i < N - j:
                return 0
            
            if (i, j) in dp:
                return dp[i, j]
            
            res = numDistinct(i + 1, j)
            
            if s[i] == t[j]:
                res+= numDistinct(i + 1, j + 1)
                 
            dp[i, j] = res
            return res
        
        return numDistinct(0, 0)